Firebase/Auth/Source/FIRAuthURLPresenter.m

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright 2017 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#import "FIRAuthURLPresenter.h"
+
+#import "FIRAuthErrorUtils.h"
+#import "FIRAuthUIDelegate.h"
+
+@import SafariServices;
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface FIRAuthDefaultUIDelegate : NSObject <FIRAuthUIDelegate>
+/** @fn defaultUIDelegate
+    @brief Returns a default FIRAuthUIDelegate object.
+    @return The default FIRAuthUIDelegate object.
+ */
++ (id<FIRAuthUIDelegate>)defaultUIDelegate;
+@end
+
+@interface FIRAuthURLPresenter () <SFSafariViewControllerDelegate>
+@end
+
+@implementation FIRAuthURLPresenter {
+  /** @var _callbackMatcher
+      @brief The callback URL matcher for the current presentation, if one is active.
+   */
+  FIRAuthURLCallbackMatcher _Nullable _callbackMatcher;
+
+  /** @var _safariViewController
+      @brief The SFSafariViewController used for the current presentation, if any.
+   */
+  SFSafariViewController *_Nullable _safariViewController;
+
+  /** @var _completion
+      @brief The completion handler for the current presentaion, if one is active.
+      @remarks This variable is also used as a flag to indicate a presentation is active.
+   */
+  FIRAuthURLPresentationCompletion _Nullable _completion;
+}
+
+- (void)presentURL:(NSURL *)URL
+        UIDelegate:(nullable id<FIRAuthUIDelegate>)UIDelegate
+   callbackMatcher:(FIRAuthURLCallbackMatcher)callbackMatcher
+        completion:(FIRAuthURLPresentationCompletion)completion {
+  _callbackMatcher = callbackMatcher;
+  _completion = completion;
+  // If a UIDelegate is not provided.
+  if (!UIDelegate) {
```

Comment:
if (!UIDelegate) {
  UIDelegate = [FIRAuthDefaultUIDelegate defaultUIDelegate];
}

Also, if we need to call dismiss on UIDelegate later, it needs to be saved as an iVar.

Reply:
done